native/dlasq2.go

```diff
@@ -0,0 +1,326 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/lapack"
+)
+
+// Dlasq2 computes all the eigenvalues of the symmetric positive
+// definite tridiagonal matrix associated with the qd array Z to high
+// relative accuracy are computed to high relative accuracy, in the
+// absence of denormalization, underflow and overflow.
+//
+// To see the relation of Z to the tridiagonal matrix, let L be a
+// unit lower bidiagonal matrix with subdiagonals Z(2,4,6,,..) and
+// let U be an upper bidiagonal matrix with 1's above and diagonal
+// Z(1,3,5,,..). The tridiagonal is L*U or, if you prefer, the
+// symmetric tridiagonal to which it is similar.
+//
+// info returns a status error.
+func (impl Implementation) Dlasq2(n int, z []float64) (info int) {
+
+	if len(z) < 4*n {
+		panic("lapack: bad z")
+	}
+	const (
+		cbias = 1.5
+	)
+
+	info = 0
+	eps := dlamchP
+	safmin := dlamchS
+	tol := eps * 100
+	tol2 := tol * tol
+	if n < 0 {
+		panic(nLT0)
+	}
+	if n == 0 {
+		return info
+	}
+	if n == 1 {
+		if z[0] < 0 {
+			panic("lapack: bad z value")
+		}
+		return info
+	}
+	if n == 2 {
+		if z[1] < 0 || z[2] < 0 {
+			panic("lapack: bad z value")
+		} else if z[2] > z[0] {
+			z[0], z[2] = z[2], z[0]
+		}
+		z[4] = z[0] + z[1] + z[2]
+		if z[1] > z[2]*tol2 {
+			t := 0.5 * (z[0] - z[2] + z[1])
+			s := z[2] * (z[1] / t)
+			if s <= t {
+				s = z[2] * (z[1] / (t * (1 + math.Sqrt(1+s/t))))
+			} else {
+				s = z[2] * (z[1] / (t + math.Sqrt(t)*math.Sqrt(t+s)))
+			}
+			t = z[0] + s + z[1]
+			z[2] *= z[0] / t
+			z[0] = t
+		}
+		z[1] = z[2]
+		z[5] = z[1] + z[0]
+		return info
+	}
+	z[2*n-1] = 0
+	emin := z[1]
+	var d, e, qmax, zmax float64
+	var i1, n1 int
+	for k := 0; k < 2*(n-1); k += 2 {
+		if z[k] < 0 {
+			panic("lapack: bad z value")
+		} else if z[k+1] < 0 {
+			panic("lapack: bad z value")
+		}
+		d += z[k]
+		e += z[k+1]
+		qmax = math.Max(qmax, z[k])
+		emin = math.Min(emin, z[k+1])
+		zmax = math.Max(math.Max(qmax, zmax), z[k+1])
+	}
+	if z[2*(n-1)] < 0 {
+		panic("lapack: bad z value")
+	}
+	d += z[2*(n-1)]
+	zmax = math.Max(qmax, zmax)
+	if e == 0 {
+		for k := 1; k < n; k++ {
+			z[k] = z[2*k]
+		}
+		impl.Dlasrt(lapack.SortDecreasing, n, z)
+		z[2*(n-1)] = d
+		return info
+	}
+	trace := d + e
+	if trace == 0 {
+		z[2*(n-1)] = 0
+		return info
+	}
+	for k := 2 * n; k >= 2; k -= 2 {
+		z[2*k-1] = 0
+		z[2*k-2] = z[k-1]
+		z[2*k-3] = 0
+		z[2*k-4] = z[k-2]
+	}
+	i0 := 0
+	n0 := n - 1
+
+	if cbias*z[4*i0] < z[4*n0] {
```

Comment:
Can you explain what's going on here with the change in indexing you've made cf the fortran?